upgrade/v122to130/upgrade.go

```diff
@@ -0,0 +1,93 @@
+package v122to130
+
+import (
+	"context"
+	"github.com/sirupsen/logrus"
+
+	"github.com/pkg/errors"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+	lhclientset "github.com/longhorn/longhorn-manager/k8s/pkg/client/clientset/versioned"
+)
+
+const (
+	upgradeLogPrefix = "upgrade from v1.2.2 to v1.3.0: "
+)
+
+func UpgradeCRs(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"UpgradeCRs failed")
+	}()
+	if err := upgradeEngines(namespace, lhClient); err != nil {
+		return err
+	}
+	return nil
+}
+
+func upgradeEngines(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "upgrade engines failed")
+	}()
+
+	engines, err := lhClient.LonghornV1beta1().Engines(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	volumeEngineMap := map[string][]*v1beta1.Engine{}
+	for i := range engines.Items {
+		e := &engines.Items[i]
+		if e.Spec.VolumeName == "" {
+			// Cannot do anything in the upgrade path if this case happens.
+			continue
+		}
+		if _, exists := volumeEngineMap[e.Spec.VolumeName]; !exists {
+			volumeEngineMap[e.Spec.VolumeName] = []*v1beta1.Engine{e}
+		} else {
+			volumeEngineMap[e.Spec.VolumeName] = append(volumeEngineMap[e.Spec.VolumeName], e)
+		}
```

Comment:
NIT: you can append to a nil slice, this will create and init a slice, so should be possible to avoid the extra comparison logic
```suggestion
		volumeEngineMap[e.Spec.VolumeName] = append(volumeEngineMap[e.Spec.VolumeName], e)
```